src/main/java/com/google/sps/DataGraph.java

```diff
@@ -184,59 +185,76 @@ public boolean mutateGraph(Mutation mut) {
+              error += "Add edge: End node " + endName + " doesn't exist";
+            }
           }
-          // The target cannot be a root since it has an in-edge
-          roots.remove(endName);
-          graph.putEdge(startNode, endNode);
           break;
         }
-      case DELETE_NODE:
+      case DELETE_EDGE:
         {
-          if (startNode == null) { // Check node exists before removing
-            return false;
-          }
-          // Check whether any successor will have no in-edges after this node is removed
-          // If so, make them roots
-
-          Set<GraphNode> successors = graph.successors(startNode);
-          for (GraphNode succ : successors) {
-            if (graph.inDegree(succ) == 1) {
-              roots.add(succ.name());
+          if (startNode != null && endNode != null) { // Check nodes exist before removing edge
+            // If the target now has no in-edges, it becomes a root
+            if (graph.inDegree(endNode) == 1) {
+              roots.add(endName);
+            }
+            graph.removeEdge(startNode, endNode);
+          } else {
+            if (startNode == null) {
+              error = "Delete edge: Start node " + startName + " doesn't exist\n";
+            }
+            if (endNode == null) {
+              error += "Delete edge: End node " + endName + " doesn't exist";
             }
           }
-          roots.remove(startName);
-          graph.removeNode(startNode); // This will remove all edges associated with startNode
-          graphNodesMap.remove(startName);
           break;
         }
-      case DELETE_EDGE:
+      case DELETE_NODE:
         {
-          if (startNode == null || endNode == null) { // Check nodes exist before removing edge
-            return false;
-          }
-          // If the target now has no in-edges, it becomes a root
-          if (graph.inDegree(endNode) == 1) {
-            roots.add(endName);
+          if (startNode != null) { // Check node exists before removing
+            // Check whether any successor will have no in-edges after this node is removed
+            // If so, make them roots
+
+            Set<GraphNode> successors = graph.successors(startNode);
+            for (GraphNode succ : successors) {
+              if (graph.inDegree(succ) == 1) {
+                roots.add(succ.name());
+              }
+            }
+            roots.remove(startName);
+            graph.removeNode(startNode); // This will remove all edges associated with startNode
+            graphNodesMap.remove(startName);
+          } else {
+            error = "Delete node: Deleting a non-existent node " + startName;
           }
-          graph.removeEdge(startNode, endNode);
           break;
         }
       case CHANGE_TOKEN:
         {
           if (startNode == null) {
-            return false;
+            error = "Change node: Changing a non-existent node " + startName;
+            break;
```

Comment:
In a PR a while back, I think Yiting suggested we do the if(condition == true) way instead of the if(!condition) way but I think then the if(!condition) branch had no associated code like setting an error. Do you think I should change it back?